src/System.Management.Automation/engine/GetCommandCommand.cs

```diff
@@ -1694,40 +1694,66 @@ private static PSObject GetParameterType(Type parameterType)
     /// </summary>
     public class NounArgumentCompleter : IArgumentCompleter
     {
+        /// </summary>
+        /// Returns completion results for Noun parameter.
+        /// </summary>
+        /// <param name="commandName">The command name.</param>
+        /// <param name="parameterName">The parameter name.</param>
+        /// <param name="wordToComplete">The word to complete.</param>
+        /// <param name="commandAst">The command AST.</param>
+        /// <param name="fakeBoundParameters">The fake bound parameters.</param>
+        /// <returns>List of completion results.</returns>
+        public IEnumerable<CompletionResult> CompleteArgument(
+            string commandName,
+            string parameterName,
+            string wordToComplete,
+            CommandAst commandAst,
+            IDictionary fakeBoundParameters) => CompletionCompleters.GetMatchingResults(
+                wordToComplete,
+                possibleCompletionValues: GetCommandNouns(fakeBoundParameters).Order());
+
         /// <summary>
+        /// Get command nouns using Get-Command.
         /// </summary>
-        public IEnumerable<CompletionResult> CompleteArgument(string commandName, string parameterName, string wordToComplete, CommandAst commandAst, IDictionary fakeBoundParameters)
+        /// <param name="fakeBoundParameters">The fake bound parameters.</param>
+        /// <returns>List of command nouns.</returns>
+        private static HashSet<string> GetCommandNouns(IDictionary fakeBoundParameters)
         {
-            if (fakeBoundParameters == null)
-            {
-                throw PSTraceSource.NewArgumentNullException(nameof(fakeBoundParameters));
-            }
+            using var ps = System.Management.Automation.PowerShell.Create(RunspaceMode.CurrentRunspace);
 
-            var commandInfo = new CmdletInfo("Get-Command", typeof(GetCommandCommand));
-            var ps = System.Management.Automation.PowerShell.Create(RunspaceMode.CurrentRunspace)
-                .AddCommand(commandInfo)
-                .AddParameter("Noun", wordToComplete + "*");
+            ps.AddCommand("Get-Command");
 
+            // Completion: Get-Command -Module <module> -Noun <wordToComplete>
             if (fakeBoundParameters.Contains("Module"))
             {
                 ps.AddParameter("Module", fakeBoundParameters["Module"]);
             }
 
-            HashSet<string> nouns = new HashSet<string>();
-            var results = ps.Invoke<CommandInfo>();
-            foreach (var result in results)
+            // Completion: Get-Command -Verb <verb> -Noun <wordToComplete>
+            if (fakeBoundParameters.Contains("Verb"))
+            {
+                ps.AddParameter("Verb", fakeBoundParameters["Verb"]);
+            }
+
+            Collection<CommandInfo> commands = ps.Invoke<CommandInfo>();
+            HashSet<string> nouns = new(commands.Count, StringComparer.OrdinalIgnoreCase);
```

Comment:
Can we use SortedSet and remove Order()?

Reply:
Ah thats a good one, always forget about SortedSet. I've updated this 🙂.

Thanks @iSazonov/